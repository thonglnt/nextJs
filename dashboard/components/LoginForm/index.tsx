import { useForm, SubmitHandler, Controller } from 'react-hook-form'
import { Box, Divider, Typography, useTheme } from '@mui/material'
import Link from 'next/link'

// Constants
import {
  INVALID_EMAIL_FORMAT,
  INVALID_PASSWORD,
  REGEX_EMAIL,
  REQUIRED,
} from '@constants/index'
import { ROUTE } from '@constants/route'

// Components
import Button from '@components/Button'
import Input from '@components/Input'
import Logo from '@components/Logo'

// Types
import { IAccount } from '@self-types/index'
// import { useRouter } from 'next/navigation'

export interface ILoginFormProps {
  handleSubmitForm: (data: IAccount) => void
  error: {
    email: string
    password: string
  }
}

const LoginForm = ({ handleSubmitForm, error }: ILoginFormProps) => {
  const theme = useTheme()
  // const router = useRouter()

  const {
    formState: { errors },
    handleSubmit,
    control,
  } = useForm<IAccount>()

  const onSubmit: SubmitHandler<IAccount> = (data) => {
    handleSubmitForm(data)
    console.log('data', data)
    // router.push(ROUTE.HOME_PAGE)
  }

  return (
    <Box
      minWidth={350}
      maxWidth={450}
      padding="32px"
      component="form"
      onSubmit={handleSubmit(onSubmit)}
      display="flex"
      justifyContent="center"
      flexDirection="column"
      alignItems="center"
      width="100%"
      bgcolor={theme.palette.common.white}
      borderRadius="8px"
      boxShadow="rgb(100 116 139 / 12%) 0px 10px 15px"
    >
      <Logo />
      <Typography
        color={theme.palette.common.black}
        fontWeight={600}
        fontSize={24}
        variant="h5"
        textAlign="center"
        lineHeight="33px"
      >
        Log in
      </Typography>
      <Typography
        color={theme.palette.grey[600]}
        fontSize={14}
        lineHeight="22px"
        marginTop="18px"
        marginBottom="25px"
      >
        Sign in on the internal platform
      </Typography>
      <Controller
        name="email"
        control={control}
        defaultValue=""
        rules={{
          required: { value: true, message: REQUIRED },
          pattern: { value: REGEX_EMAIL, message: INVALID_EMAIL_FORMAT },
        }}
        render={({ field }) => (
          <Input
            label="Email Address"
            error={!!errors.email || !!error.email}
            errorMsg={
              ((errors.email && errors.email.message) || error.email) as string
            }
            type="text"
            fullWidth
            {...field}
          />
        )}
      />
      <Controller
        name="password"
        control={control}
        defaultValue=""
        rules={{
          required: { value: true, message: REQUIRED },
          minLength: { value: 8, message: INVALID_PASSWORD },
        }}
        render={({ field }) => (
          <Input
            label="Password"
            error={!!errors.password || !!error.password}
            errorMsg={
              ((errors.password && errors.password.message) ||
                error.password) as string
            }
            type="password"
            fullWidth
            {...field}
            sx={{
              marginTop: '15px',
              marginBottom: '25px',
            }}
          />
        )}
      />
      <Button type="submit" fullWidth variant="contained" color="secondary">
        Log In
      </Button>
      <Divider
        sx={{
          width: '100%',
          height: '1px',
          marginTop: '25px',
          marginBottom: '30px',
        }}
      />
      <Box
        display="flex"
        flexDirection="column"
        gap={1}
        alignItems="flex-start"
      >
        <Link href={ROUTE.REGISTER} prefetch={false}>
          <Typography
            paddingLeft={0.5}
            color={theme.palette.info.main}
            fontSize={14}
            fontWeight={500}
          >
            Create new account
          </Typography>
        </Link>
        <Link href={ROUTE.FORGOT_PASSWORD} prefetch={false}>
          <Typography
            paddingLeft={0.5}
            color={theme.palette.info.main}
            fontSize={14}
            fontWeight={500}
          >
            Forgot password
          </Typography>
        </Link>
      </Box>
    </Box>
  )
}

export default LoginForm
