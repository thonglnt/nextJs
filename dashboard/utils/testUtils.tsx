import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'

// Themes
import theme from '@self-types/Themes.types'

export const customRender = (children: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
}

export * from '@testing-library/react'
export { customRender as render }
