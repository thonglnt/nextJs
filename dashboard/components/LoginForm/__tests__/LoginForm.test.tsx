// Utils
import { fireEvent, render, screen } from '@utils/testUtils'

// Components
import LoginForm from '..'

describe('Login Form component', () => {
  const handleSubmit = jest.fn()
  const props = {
    email: 'thuandang@gmail.com',
    password: '123456789',
  }
  it('should render Login form', () => {
    const { container } = render(
      <LoginForm
        handleSubmitForm={handleSubmit}
        error={{
          email: '',
          password: '',
        }}
      />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should call onChange input email', () => {
    render(
      <LoginForm
        handleSubmitForm={handleSubmit}
        error={{
          email: '',
          password: '',
        }}
      />,
    )
    const input = screen.getByRole('textbox')

    fireEvent.change(input, {
      target: { value: props.email },
    })

    expect(input).toHaveValue(props.email)
  })

  it('should call onChange input password', () => {
    render(
      <LoginForm
        handleSubmitForm={handleSubmit}
        error={{
          email: '',
          password: '',
        }}
      />,
    )
    const input = screen.getByRole('textbox')

    fireEvent.change(input, {
      target: { value: props.password },
    })

    expect(input).toHaveValue(props.password)
  })
})
