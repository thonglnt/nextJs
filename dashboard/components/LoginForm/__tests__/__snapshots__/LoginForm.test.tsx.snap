// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form component should render Login form 1`] = `
<div>
  <form
    class="MuiBox-root css-dnnnuz"
  >
    <a
      href="/"
      style="max-width: 40px; max-height: 40px;"
    >
      <img
        alt="logo"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="eager"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=40"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=40 1x, /_next/image?url=%2Fimg.jpg&w=96&q=40 2x"
        style="color: transparent;"
        width="40"
      />
    </a>
    <h5
      class="MuiTypography-root MuiTypography-h5 css-13ikgl9-MuiTypography-root"
    >
      Log in
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body1 css-bzfmi6-MuiTypography-root"
    >
      Sign in on the internal platform
    </p>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-xne6jv-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-qmlsft-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="outlined-required"
        id="outlined-required-label"
      >
        Email Address
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl css-1cswynb-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-required"
          name="email"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Email Address
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-9tsqn3-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-qmlsft-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="outlined-required"
        id="outlined-required-label"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl css-1cswynb-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-required"
          name="password"
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1marung-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Log In
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-1l74i5m-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-bgm2id"
    >
      <a
        href="/register"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ai1txx-MuiTypography-root"
        >
          Create new account
        </p>
      </a>
      <a
        href="/forgot-password"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ai1txx-MuiTypography-root"
        >
          Forgot password
        </p>
      </a>
    </div>
  </form>
</div>
`;
